Set up CI on Windows

lishaduck opened the pull request:
Resolves #343.

~~Should probably fail until #340 gets fixed?~~

lishaduck committed:
ci: setup ci on windows

jfmengels commented:
I agree we should run tests on Windows (maybe on Mac as well?), but I imagine it will increase the likelyhood that we get rate limits (which we still keep seeking for some reason). Should we maybe exclude the `template` tests on Windows?

lishaduck commented:
First off, how did I hit rate limits without running it? Must've still got something wrong with #310. I'll take a look ASAP.
Second off, it luckily looks like most of this so far is newlines. A few path issues _in test code_, but I didn't see any bugs (which are probably in the run tests).

lishaduck committed:
ci: less windows

elm-format doesn't look like it supports lf line endings on Windows.
Also disable remote tests on Windows.

lishaduck committed:
ci: replace CRLF with LF on Windows

lishaduck committed:
wip

I think this is partly a step in the wrong direction, but we'll see.

Here's what I've found about drive letters:
- On normal Windows, paths are DOS paths, like "C:\path\to\file"
- Except when they're not, like
  - "\\.\C:\path\to\file"
  - "\\?\C:\path\to\file"
- Sometimes Windows uses UNC, which is like "\\hostname\C$\path\to\file"
- On Git Bash, paths are like "/c/path/to/file"
- On Node, paths are like "C:/path/to/file"
- On WSL, paths are like "/mnt/c/path/to/file"

Sometimes drive letters are case sensitive, sometimes they're not.

And that's just for absolute paths.
There's even more weird stuff for relative paths.
Makes me glad to be using macOS so much these days.

lishaduck committed:
ci: better matrix

lishaduck committed:
ci: try pathe

lishaduck committed:
fix: shhhh lint

Yes, this should be fixed.
No, this should be fixed later.
For now, I need to check the tests.

lishaduck committed:
ci: still run engines test

lishaduck commented on test/run.mjs:
```diff
@@ -48,7 +48,13 @@ if (nodeVersion !== expectedVersion) {
-  return data.replace(new RegExp(localPath, 'g'), '<local-path>');
+  return data
+    .replace(
+      // eslint-disable-next-line security/detect-non-literal-regexp -- Test code.
```
I'm not happy about this either. Perhaps we should only enable security on non-test files?

jfmengels replied:
Depends what other rules `security` holds. I would like it help prevent if people try to do weird code executions for instance. But yeah, a regex like this is fine.

lishaduck commented:
Wow, pathe is magic! All tests failing to just one failing because Anonymize needs updates.

lishaduck committed:
do we still need this?

lishaduck committed:
chore: add comment

jfmengels commented on lib/watch.js:
```diff
@@ -383,18 +376,15 @@ function createExtraFilesWatcher(options, app, runReview, onError, request) {
       request.files.map(
-        ({pattern, included}) =>
-          (included ? '' : '!') + OsHelpers.makePathOsAgnostic(pattern)
+        ({pattern, included}) => (included ? '' : '!') + pattern
```
Does chokidar support Windows-style paths?

lishaduck replied:
~~No idea. I should check.~~ They normalize it (at least they used to, we're still on v3, so they still should?), so you can use either, but it _should_ return a POSIX path rather than a DOS path.
I had a chat with my dad about the elm-review's handling of drives, which basically ended with `makePathOsAgnostic` uses `/.:/` to strip the drive, when elm-review should keep the drive so that you can execute it across drives (which MS wants to make more common). `pathe` handles proper agnosticism, so the pattern passed in should be correct, but I should also just check that as well.

lishaduck committed:
fix: don't use pathe in anonymize

lishaduck commented:
@jfmengels, how much do we care about outputting idiomic paths on Windows? Right now, the CLI outputs paths like this: "C:\Somewhere\Foo.txt", but this PR changes that to "C:/Somewhere/Foo.txt", which is _equally valid_, but could be a slight behavioral change (does that hurt the IDE integration?) and might be confusing to newcomers? (But Windows paths are confusing in general, so maybe they should just get over it, see the writeup at https://github.com/jfmengels/node-elm-review/pull/344/commits/b9c822a1113ac15879965b21d53f3db31bd6df9d).

lishaduck committed:
fix: anonymize more

lishaduck committed:
fix: normalizer

lishaduck commented:
@jfmengels, can you take a look a the current failure? I'm not really sure why it's different. Is it an prompts thing? I don't think we're doing anything different.

lishaduck commented:
> Is it an prompts thing?

Yup. This is silly code: https://github.com/terkelg/prompts/blob/e0519913ec4fcc6746bb3d97d8cd0960c3f3ffde/lib/util/figures.js#L24

lishaduck commented:
> Can Windows really not render the characters by default?

If I'm reading [this old DevBlogs post](https://devblogs.microsoft.com/commandline/windows-command-line-unicode-and-utf-8-output-text-buffer/) correctly, UTF-8 support for Windows Console landed in 2018, so it should be fine by now.

jfmengels commented:
> how much do we care about outputting idiomic paths on Windows? Right now, the CLI outputs paths like this: "C:\Somewhere\Foo.txt", but this PR changes that to "C:/Somewhere/Foo.txt", which is equally valid, but could be a slight behavioral change (does that hurt the IDE integration?) and might be confusing to newcomers? (But Windows paths are confusing in general, so maybe they should just get over it, see the writeup at https://github.com/jfmengels/node-elm-review/commit/b9c822a1113ac15879965b21d53f3db31bd6df9d).

That's a **lot** of ways to write Windows paths :thinking: 
I honestly don't know. If they're bot h valid, then either one is fine by me, but since I'm not a Windows dev, I'm not adequately positioned to to tell what is best. I would say that it's probably best to change this only it it is nicer for the user in some way, not if it's just for us running it in CI.

lishaduck commented:
> I would be okay with replacing `prompts` with a custom thing as well, but that's a bigger task 😅

I took a closer look at alternatives. Clack is a very different feel from elm-review's current vibe, I don't think we want to go that way. inquirer also detects windows, but it also supports unicode on modern terminals, so perhaps we could fake the terminal env var in tests to be one that supports unicode? Maybe???
I might need to look into the prior art. Somebody's got to have run into this before.

lishaduck commented:
Huh. This is weird, I just realized pathe v2 is ESM-only, but we're importing it in CJS without issues (and on v16, not like v23 where that's fine)

lishaduck commented:
I've been busy with life (still am, hopefully should slow up soon), but have been thinking about this.
1. Let's just find+replace this for now
2. > Clack is a very different feel from elm-review's current vibe

    I've looked into clack some more in general, and while I'd like to use it, it's both ESM-only and would take a lot of effort to make it look like elm-review.

lishaduck committed:
chore: document normalize function

lishaduck committed:
revert: drop pathe

We want to keep giving normally-styled Windows paths in errors.
pathe doesn't let us do that easily.

lishaduck committed:
wip: replace prompts chars

lishaduck committed:
fix: pathe in test/ again

lishaduck committed:
fix: more pathe in test/

lishaduck committed:
test: replace windows paths

lishaduck committed:
test: replace more windows paths

lishaduck committed:
test: replace drives better

lishaduck committed:
fix: json diff

This is getting really hacky.

lishaduck committed:
fix: use regex

replace isn't replaceAll.

lishaduck committed:
refactor: use process.versions.node

lishaduck committed:
test: even hackier

lishaduck committed:
test: even hackier

lishaduck committed:
refactor: centralize anonymization

lishaduck committed:
refactor: don't anonmyize windows stuff all the time

lishaduck committed:
refactor: don't anonymize versions as much

lishaduck commented:
This PR's diff is going to be such a mess.

lishaduck committed:
fix: Y/n

lishaduck committed:
test: replace json quotes in strings right

This is getting really stupid.

lishaduck committed:
fix: use a more specific regex

"Back in the habit!"

lishaduck committed:
fix: cp

lishaduck commented:
Was looking through the errors and was trying to debug an issue... and it turns out it's 5 years old 😂 
https://github.com/jfmengels/node-elm-review/commit/39ec4e5f5fc36d6c5e3c80d635738771f8132229#diff-9218c1cd9409fa9cb2e842d38893e5e087b82501b9a6accc052ce90c3170cf9b

lishaduck committed:
wip

lishaduck committed:
fix: patch prompts in tests???

This is getting to be such a pain...

lishaduck committed:
fix: old local patch

lishaduck committed:
tada

jfmengels commented:
This looks great! Great work! Some tests are still failing, for different reasons in both OSes.

lishaduck commented:
> This looks great! Great work! Some tests are still failing, for different reasons in both OSes.

Yeah, it's not quite done but I think I ported over most of the less invasive changed from the `windows-2` branch + I improved the regexes from this branch & so excepting actual bugs on Windows, I think it shouldn't be too bad to finish.
I'll be on Winter break from school, so I should hopefully be able to find some time to wrap this up soon.

lishaduck committed:
fix: tests on unixy

Hey, this should help minimize the breakage when we release...

---

Also:

chatgpt is the ruination of society...
but it sure is better than me at regex, and I can't use elm/parser so :)

aka explainable semi-stupidity/semi-overly-cleverness

lishaduck committed:
chore: fewer random spaces

Could've sworn I'd removed these already.

lishaduck committed:
fix: zx on windows

I thought the whole point was so I didn't have to write all this out... clearly not.

lishaduck committed:
fix: make git checkout in jest go shhh 🤫

lishaduck commented on patches/prompts+2.4.2.patch:
```diff
@@ -0,0 +1,12 @@
+   pointer: '>'	
+ };	
+-const figures = process.platform === 'win32' ? win : main;	
++const figures = main;	
```
> > And this is only used locally, right?
> 
> Yup, it runs in prepare so it only runs when we install, not when users install.

...which turns out to be an issue cuz half of the tests this targets are integration tests that don't pick up our node_modules (on purpose)...
Thus, reverting this, unfortunately. Back to regexes...

lishaduck committed:
Revert "fix: patch prompts in tests???"

This reverts commit 8c810eb462179d0d61aebbd351f2fdd8b7dd9426.

---

Also vibe back os agnostic paths...
To be manually verified on a real project at a later time, but tests pass and eyeing it, seems correct.

lishaduck commented on lib/error-message.js:
```diff
@@ -29,19 +29,19 @@ class CustomError extends Error {
+  return Anonymize.pathsAndVersions(
+    formatHuman(options.debug, err),
+    options.forTests
+  );
```
Note to self: reevalute anonymizing in this function. It's nice to have it, but we end up needing to normalize elsewhere anyway, so perhaps we could simplify this logic by simply removing it.

lishaduck replied:
> I'm guessing we can do a find+replace in the output in a helper from our test suite.

Yeah, we have to do that anyway because this doesn't fully anonymize all error paths, so I figured we might as well drop this (which simplifies some of the windows logic).

lishaduck committed:
maybe???

Co-authored-by: Brian Dukes <bdukes@engagesoftware.com>

lishaduck committed:
fix: npx on windows

Why this breaks, no idea.
Probably a new node issue
(running v20 locally
b/c it's the first version to support Windows ARM64).

lishaduck committed:
debugging

lishaduck committed:
chore: update elm-tooling to support win-arm64
